### ISSUE TITLE
Websocket does not work

### ISSUE THREAD

**Issue by dikderoy** [TykTechnologies/tyk#1339]
**Do you want to request a *feature* or report a *bug*?**
Bug
**What is the current behavior?**
When opening a ws connection following exception is raised in Tyk Gateway logs:
```
tyk_gateway_1    | 2017/12/04 05:24:32 http: panic serving 172.18.0.18:53856: interface conversion: http.RoundTripper is *main.WSDialer, not *http.Transport
tyk_gateway_1    | goroutine 43779 [running]:
tyk_gateway_1    | net/http.(*conn).serve.func1(0xc4201d5400)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1721 +0xd0
tyk_gateway_1    | panic(0xcfedc0, 0xc42041ca40)
tyk_gateway_1    | 	/usr/local/go/src/runtime/panic.go:489 +0x2cf
tyk_gateway_1    | main.(*ReverseProxy).WrappedServeHTTP(0xc4203e3360, 0x1330ee0, 0xc42023b260, 0xc420069a00, 0x5a24dc00, 0x0)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/reverse_proxy.go:524 +0x1d0f
tyk_gateway_1    | main.(*ReverseProxy).ServeHTTP(0xc4203e3360, 0x1330ee0, 0xc42023b260, 0xc420069a00, 0x120)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/reverse_proxy.go:349 +0x60
tyk_gateway_1    | main.(*SuccessHandler).ServeHTTP(0xc420249e00, 0x1330ee0, 0xc42023b260, 0xc420069a00, 0x31048f40)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/handler_success.go:218 +0x16e
tyk_gateway_1    | main.(*DummyProxyHandler).ServeHTTP(0xc420249e00, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/api_loader.go:518 +0x4b
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199380, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc4201993c0, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199400, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199440, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199480, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc4201994c0, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199500, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199540, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc420199580, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | main.createMiddleware.func1.1(0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/middleware.go:88 +0x7ec
tyk_gateway_1    | net/http.HandlerFunc.ServeHTTP(0xc4201995c0, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1942 +0x44
tyk_gateway_1    | github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux.(*Router).ServeHTTP(0xc4202568c0, 0x1330ee0, 0xc42023b260, 0xc420069a00)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/vendor/github.com/gorilla/mux/mux.go:114 +0x10c
tyk_gateway_1    | main.mainHandler.ServeHTTP(0x1330ee0, 0xc42023b260, 0xc420069800)
tyk_gateway_1    | 	/src/github.com/TykTechnologies/tyk/main.go:1272 +0x4b
tyk_gateway_1    | main.(*mainHandler).ServeHTTP(0x142a620, 0x1330ee0, 0xc42023b260, 0xc420069800)
tyk_gateway_1    | 	<autogenerated>:744 +0x63
tyk_gateway_1    | net/http.serverHandler.ServeHTTP(0xc420236210, 0x1330ee0, 0xc42023b260, 0xc420069800)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:2568 +0x92
tyk_gateway_1    | net/http.(*conn).serve(0xc4201d5400, 0x1331d20, 0xc42018fe00)
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:1825 +0x612
tyk_gateway_1    | created by net/http.(*Server).Serve
tyk_gateway_1    | 	/usr/local/go/src/net/http/server.go:2668 +0x2ce
```

**What is the expected behavior?**
to proxy websocket connection normally
**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem**
an nginx facing a client and proxying request to tyk_gateway with
configuration as following:

```
map $http_upgrade $connection_upgrade {
    default upgrade;
    ""      close;
}

upstream tyk_gateway {
    server tyk_gateway:8080;
}

server {
    listen 80;
    server_name example.local;

    root /usr/share/nginx/html;

    location /api/ {
        location ~ /workspaces/[A-Za-z0-9-]+/chat {
            rewrite /api/(.*) /$1  break;
            proxy_redirect     off;

            proxy_pass http://tyk_gateway;
            proxy_http_version 1.1;
            proxy_set_header Upgrade $http_upgrade;
            proxy_set_header Connection $connection_upgrade;
            proxy_read_timeout 600s;
        }

        proxy_pass http://tyk_gateway/;
    }
}
```

tyk configuration:

```json
{
    "listen_port": 8080,
    "secret": "***",
    "node_secret": "***",
    "template_path": "/opt/tyk-gateway/templates",
    "tyk_js_path": "/opt/tyk-gateway/js/tyk.js",
    "middleware_path": "/opt/tyk-gateway/middleware",
    "use_db_app_configs": true,
    "db_app_conf_options": {
        "connection_string": "http://tyk_dashboard:3000",
        "node_is_segmented": false,
        "tags": ["test2"]
    },
    "app_path": "/opt/tyk-gateway/apps/",
    "storage": {
        "type": "redis",
        "host": "redis",
        "port": 6379,
        "username": "",
        "password": "",
        "database": 0,
        "optimisation_max_idle": 100
    },
    "http_server_options":{
        "enable_websockets": true
    },
    "enable_analytics": true,
    "analytics_config": {
        "type": "mongo",
        "csv_dir": "/tmp",
        "mongo_url": "",
        "mongo_db_name": "",
        "mongo_collection": "",
        "purge_delay": -1,
        "ignored_ips": []
    },
    "health_check": {
        "enable_health_checks": true,
        "health_check_value_timeouts": 60
    },
    "optimisations_use_async_session_write": true,
    "enable_non_transactional_rate_limiter": true,
    "enable_sentinel_rate_limiter": false,
    "allow_master_keys": false,
    "policies": {
        "policy_source": "service",
        "policy_connection_string": "http://tyk_dashboard:3000",
        "policy_record_name": "tyk_policies"
    },
    "hash_keys": true,
    "close_connections": true,
    "allow_insecure_configs": true,
    "coprocess_options": {
        "enable_coprocess": false,
        "coprocess_grpc_server": ""
    },
    "enable_bundle_downloader": true,
    "bundle_base_url": "",
    "global_session_lifetime": 100,
    "force_global_session_lifetime": false,
    "max_idle_connections_per_host": 100
}
```
api config:

```json
{
  "id": "5a215fbe4665480001aeb34b",
  "name": "chat_ws",
  "slug": "chat_ws",
  "api_id": "a88b54edd45e484e502ed7d0e6033a08",
  "org_id": "5a1f2980214f47000162eb93",
  "use_keyless": false,
  "use_oauth2": false,
  "use_openid": false,
  "openid_options": {
    "providers": [],
    "segregate_by_client": false
  },
  "oauth_meta": {
    "allowed_access_types": [],
    "allowed_authorize_types": [],
    "auth_login_redirect": ""
  },
  "auth": {
    "use_param": false,
    "param_name": "",
    "use_cookie": true,
    "cookie_name": "_reason_token",
    "auth_header_name": "Authorization",
    "use_certificate": false
  },
  "use_basic_auth": false,
  "use_mutual_tls_auth": false,
  "client_certificates": [],
  "upstream_certificates": {},
  "enable_jwt": false,
  "use_standard_auth": true,
  "enable_coprocess_auth": false,
  "jwt_signing_method": "",
  "jwt_source": "",
  "jwt_identity_base_field": "",
  "jwt_client_base_field": "",
  "jwt_policy_field_name": "",
  "notifications": {
    "shared_secret": "",
    "oauth_on_keychange_url": ""
  },
  "enable_signature_checking": false,
  "hmac_allowed_clock_skew": -1,
  "base_identity_provided_by": "",
  "definition": {
    "location": "header",
    "key": "x-api-version"
  },
  "version_data": {
    "not_versioned": true,
    "versions": {
      "Default": {
        "name": "Default",
        "expires": "",
        "paths": {
          "ignored": [],
          "white_list": [],
          "black_list": []
        },
        "use_extended_paths": true,
        "extended_paths": {},
        "global_headers": {
          "X-User-Id": "$tyk_meta.tyk_developer_id"
        },
        "global_headers_remove": [],
        "global_size_limit": 0,
        "override_target": ""
      }
    }
  },
  "uptime_tests": {
    "check_list": [],
    "config": {
      "expire_utime_after": 0,
      "service_discovery": {
        "use_discovery_service": false,
        "query_endpoint": "",
        "use_nested_query": false,
        "parent_data_path": "",
        "data_path": "",
        "port_data_path": "",
        "target_path": "",
        "use_target_list": false,
        "cache_timeout": 60,
        "endpoint_returns_list": false
      },
      "recheck_wait": 0
    }
  },
  "proxy": {
    "preserve_host_header": false,
    "listen_path": "/workspaces/{id:.*}/chat",
    "target_url": "ws://ws_api:8080/",
    "strip_listen_path": false,
    "enable_load_balancing": false,
    "target_list": [],
    "check_host_against_uptime_tests": false,
    "service_discovery": {
      "use_discovery_service": false,
      "query_endpoint": "",
      "use_nested_query": false,
      "parent_data_path": "",
      "data_path": "",
      "port_data_path": "",
      "target_path": "",
      "use_target_list": false,
      "cache_timeout": 0,
      "endpoint_returns_list": false
    }
  },
  "disable_rate_limit": false,
  "disable_quota": false,
  "custom_middleware": {
    "pre": [],
    "post": [],
    "post_key_auth": [],
    "auth_check": {
      "name": "",
      "path": "",
      "require_session": false
    },
    "response": [],
    "driver": "",
    "id_extractor": {
      "extract_from": "",
      "extract_with": "",
      "extractor_config": {}
    }
  },
  "custom_middleware_bundle": "",
  "cache_options": {
    "cache_timeout": 60,
    "enable_cache": false,
    "cache_all_safe_requests": false,
    "cache_response_codes": [],
    "enable_upstream_cache_control": false
  },
  "session_lifetime": 0,
  "active": true,
  "auth_provider": {
    "name": "",
    "storage_engine": "",
    "meta": {}
  },
  "session_provider": {
    "name": "",
    "storage_engine": "",
    "meta": {}
  },
  "event_handlers": {
    "events": {}
  },
  "enable_batch_request_support": false,
  "enable_ip_whitelisting": false,
  "allowed_ips": [],
  "dont_set_quota_on_create": false,
  "expire_analytics_after": 0,
  "response_processors": [],
  "CORS": {
    "enable": false,
    "allowed_origins": [],
    "allowed_methods": [],
    "allowed_headers": [],
    "exposed_headers": [],
    "allow_credentials": false,
    "max_age": 24,
    "options_passthrough": false,
    "debug": false
  },
  "domain": "",
  "do_not_track": false,
  "tags": [],
  "enable_context_vars": false,
  "config_data": {},
  "tag_headers": [],
  "global_rate_limit": {
    "rate": -1,
    "per": -1
  },
  "strip_auth_data": false
}
```

and a python aiohttp websocket app behind tyk gateway

requesting websocket connection from JS using native API

```javascript
        var sock_addr = '/api/workspaces/' + workspaces.val() + '/chat';
        try {
            sock = new WebSocket('ws://' + window.location.host + sock_addr);
            console.log('socket opened')
        }
        catch (err) {
            alert('failed to open ws')
        }
```

connection is actually "closes" right away (but this is because it is established just fine with nginx)

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**
2.4.0, 2.4.1 (docker image build)

**Comment by buger** [comment]
Thank you for the report! It is definitely a regression.

**Comment by dencoded** [comment]
Just to follow up - I've looked at code and found a potential problem there. I will provide fix asap.

**Comment by dikderoy** [comment]
Is there any way to make websockets work without the fix? or are they rendered completely broken?

I have tried to arrange same without nginx and on tyk 2.3.13 and it also does not work (same error). maybe there is something wrong with the config/definition?

**Comment by dencoded** [comment]
Hey @dikderoy 
Thanks again for reporting about this problem and for your patience. I pushed PR (see above) which I believe should fix this problem.

**Comment by buger** [comment]
Fixed with 2.4.2 👌

**Comment by dikderoy** [comment]
I've checked the fix (in 2.4.2) - initially websocket connection is working, but as soon as you drop "first" connection, it stops working, and you cannot create new ws-connection without restarting tyk_gateway.

tyk gateway produces 500 Internal Server Error on ws.open() with following logs:

```
tyk_gateway_1    | time="Dec 22 15:57:10" level=error msg="Hijack error: http: connection has been hijacked" origin=172.18.0.18 path="/workspaces/default/chat"
tyk_gateway_1    | time="Dec 22 15:57:10" level=error msg="http: proxy error: http: connection has been hijacked" api_id=b7c7ce1ff3f449357eb28312a6fe8fe7 org_id=5a2b3da0bdf125000143add4 server_name="ws_api:8080" user_id="****b567" user_ip=172.18.0.18 user_name="dik.deroy@gmail.com (platform)"
```
